lib/truffle/stringio.rb

```diff
@@ -280,6 +280,13 @@ def write(str)
     str = String(str)
     return 0 if str.empty?
 
+    if external_encoding &&
+       external_encoding != str.encoding &&
+       external_encoding != Encoding::BINARY &&
+       str.encoding != Encoding::BINARY
+      str = str.encode(external_encoding)
+    end
```

Comment:
We have the same logic in `IO#write`, it makes sense to share it to avoid duplication and keep the logic consistent. I'll move that logic to `Truffle::IOOperations`.